catboost/libs/metrics/metric.cpp

```diff
@@ -466,6 +467,74 @@ void TMultiRMSEMetric::GetBestValue(EMetricBestValue* valueType, float* /*bestVa
     *valueType = EMetricBestValue::Min;
 }
 
+/* MultiRMSEWithMissingValues */
+namespace {
+    struct TMultiRMSEWithMissingValues final: public TAdditiveMultiRegressionMetric {
+        explicit TMultiRMSEWithMissingValues(const TLossParams& params)
+            : TAdditiveMultiRegressionMetric(ELossFunction::MultiRMSE, params)
+        {}
+
+        static TVector<THolder<IMetric>> Create(const TMetricConfig& config);
+
+        TMetricHolder EvalSingleThread(
+            TConstArrayRef<TVector<double>> approx,
+            TConstArrayRef<TVector<double>> approxDelta,
+            TConstArrayRef<TConstArrayRef<float>> target,
+            TConstArrayRef<float> weight,
+            int begin,
+            int end
+        ) const override;
+        void GetBestValue(EMetricBestValue* valueType, float* bestValue) const override;
+        double GetFinalError(const TMetricHolder& error) const override;
+    };
+}
+
+// static
+TVector<THolder<IMetric>> TMultiRMSEWithMissingValues::Create(const TMetricConfig& config) {
+    return AsVector(MakeHolder<TMultiRMSEWithMissingValues>(config.Params));
+}
+
+TMetricHolder TMultiRMSEWithMissingValues::EvalSingleThread(
+    TConstArrayRef<TVector<double>> approx,
+    TConstArrayRef<TVector<double>> approxDelta,
+    TConstArrayRef<TConstArrayRef<float>> target,
+    TConstArrayRef<float> weight,
+    int begin,
+    int end
+) const {
+    const auto evalImpl = [=](bool useWeights, bool hasDelta) {
+        const auto realApprox = [=](int dim, int idx) { return approx[dim][idx] + (hasDelta ? approxDelta[dim][idx] : 0); };
+        const auto realWeight = [=](int idx) { return useWeights ? weight[idx] : 1; };
+
+        TMetricHolder error(1);
+        for (auto dim : xrange(target.size())) {
+            double w = 0.f;
+            double e = 0.f;
+            for (auto i : xrange(begin, end)) {
+                if (std::isnan((float)target[dim][i])) {
+                    continue;
+                } else {
+                    e += realWeight(i) * Sqr(realApprox(dim, i) - target[dim][i]);
+                    w += realWeight(i);
+                }
+            }
+            error.Stats[0] += e/w;
```

Comment:
It isn't correct because `EvalSingleThread` can be called several times for batches of dataset and then `GetFinalError` will be called. You should add `sumErrors` to `Stats[dim*2]` and `sumWeights` to `Stats[dim*2+1]`. Then in the `GetFinalError` you should divide `Stats[dim*2]` by `Stats[dim*2+1]` for each dimension and sum all results.

Reply:
Fixed.